content/blog/2019-09-25-type_checker.md

```diff
@@ -0,0 +1,88 @@
+[Neil Adit](http://www.neiladit.com) is (guess what!) also a 2nd year PhD student at CSL, Cornell. He is interested in fun things in compiler research.
+
+"""
+extra.latex = true
++++
+
+
+
+## Goal
+
+The goal of the project was to add a static type checker to find type errors, multiple definitions of variables and undefined ones as well. We also ensure that the branch labels are valid and unique.
+
+
+
+## Design and Implementation
+
+### Design
+
+Bril currently supports 2 types - **int** and **bool** which makes type checking relatively easy. Our type checker is defined such that an arithmetic operation  `a = b+c` would raise an error if either the source or destination arguments are *not* int. Similarly, boolean operations only accept all bool arguments and comparison operations have integer arguments but a boolean destination. These type definitions have been nicely defined [here](https://capra.cs.cornell.edu/bril/langref.html). Currently Bril only supports operations on variables and not constants which implies `a = b+5` is an invalid operation. During type checking, we also ensure that the variable has been defined in the function before. Conversely, we make sure that there isn't a redefinition of the variable. For control flow operation, we ensure that labels are present in the code and are uniquely defined. This allows us to treat label strings as a separate type, invisible to the user. 
+
+The error raised by the type checker outputs the line at which we find the first operation breaking the type check rule and mentions the type of error found.
+
+
+### Implementation
+
+We have a first pass in the algorithm which collects a list of labels and ensures that each label is unique. If there are multiple labels with the same name, we throw an error at this point. This also helps us create a set of label names which is used to check for valid strings (of labels) during the second pass for control flow operations.  In the second pass, we go over each instruction and check for various type errors. Some fundamental checks in this process are:
+
+1. Invalid instructions- This ensures that all the arguments and destination (if applicable) of an instruction are available, which is given by the length of instruction.
+2. Argument and destination type- For various operations we check if arguments and destination variables have the correct type. Something like `bool d = a + b` where `a,b` are integers would raise an error for the destination variable.
+3. Redefined variables- We check if the destination variable has already been assigned to a different type in that context. We do allow const instructions on the same type as it can be treated as an assignment operation rather than a definition. So `int a =2; bool a = true` is not allowed but `int a =2; int a =5` is allowed. We do this by keeping a set of variables of each type (int and bool) defined in the function. This helps while checking existing definitions and possible redefinition errors.
```

Comment:
> We do allow const instructions on the same type as it can be treated as an assignment operation rather than a definition.

Is this something special about "const," or is it true of all instructions? It seems to be implying the former, but I think it's the latter.